### PR TITLE
[feature] Tagged logging

### PR DESCRIPTION
**Pull request by mike-north**
> **PLEASE MERGE #6 BEFORE THIS**

This feature allows prefixes to be pushed onto (and popped off of) a stack, accomplishing something similar to how [Rails ActiveSupport/TaggedLogging](https://api.rubyonrails.org/classes/ActiveSupport/TaggedLogging.html) works

 ```ts
const logger = new Logger();

 function fooFunction() {
   logger.pushPrefix('foo');
   ...
   logger.log('hello');
   logger.log('world');
   ...
    logger.popPrefix();
 }

 logger.log('begin');   // "begin"
 fooFunction();         // "[foo]: hello"
                        // "[foo]: world"
 logger.log('end');     // "end"
```

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -63,6 +68,48 @@ class Logger {
...
+   *   logger.pushPrefix('foo');
+   *   ...
+   *   logger.log('hello');
+   *   logger.log('world');
+   *   ...
+   *    logger.popPrefix();
+   * }
+   *
+   * logger.log('begin');   // "begin"
+   * fooFunction();         // "[foo]: hello"
```
Comment (lisaychuang) [inline src/index.ts]: Can we eliminate the colons?

File: test/tagged-logging-test.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    const logger = new Logger(4, printer);
+    logger.log('hello');          // "hello"
+    logger.log('world');          // "world"
+    logger.pushPrefix('foo');
+    logger.log('goodbye');        // "[foo]: goodbye"
+    logger.log('mars');           // "[foo]: mars"
+    logger.pushPrefix('bar');
+    logger.log('(and saturn!)');  // "[foo][bar]: (and saturn)!"
+    logger.popPrefix();
+    logger.popPrefix();
```
Comment (lisaychuang) [inline test/tagged-logging-test.ts]: What happens if the array is empty and there are no more Prefixes?

Comment (mike-north) [reply test/tagged-logging-test.ts]: I have confirmed that it will not be harmful, and consistent with what developers expect from `[].pop();`

**Comment by lisaychuang** [comment]
You work fast! Thank you for the PR :shipit:

**Pull request merged**